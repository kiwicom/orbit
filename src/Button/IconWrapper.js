// @flow
import * as React from "react";
import * as tokens from "@kiwicom/orbit-design-tokens";

const sizeIcon = {
  small: tokens.heightIconSmall,
  normal: tokens.heightIconMedium,
  large: tokens.heightIconMedium,
};

const marginRight = {
  small: tokens.marginButtonIconSmall,
  normal: tokens.marginButtonIconNormal,
  large: tokens.marginButtonIconLarge,
};

const colorIcon = {
  primary: tokens.colorIconButtonPrimary,
  secondary: tokens.colorIconButtonSecondary,
  link: tokens.colorIconButtonLink,
  facebook: tokens.colorIconButtonFacebook,
  google: tokens.colorIconButtonGoogle,
  destructive: tokens.colorIconButtonDestructive,
};

const colorIconBordered = {
  primary: tokens.colorIconButtonPrimaryBordered,
  secondary: tokens.colorIconButtonSecondaryBordered,
  link: tokens.colorIconButtonLinkBordered,
  facebook: tokens.colorIconButtonFacebookBordered,
  google: tokens.colorIconButtonGoogleBordered,
  destructive: tokens.colorIconButtonDestructiveBordered,
};

type Props = {
  bordered?: boolean,
  size: $Keys<typeof sizeIcon>,
  type: $Keys<typeof colorIcon> | $Keys<typeof colorIconBordered>,
  Icon: React.ComponentType<*>,
};

const IconWrapper = (props: Props) => {
  const { Icon, type, size, bordered } = props;

  return (
    <React.Fragment>
      <Icon color={bordered ? colorIconBordered[type] : colorIcon[type]} size={sizeIcon[size]} />
      <style xml>{`
        svg {
          margin-right: ${marginRight[props.size]};
        }
      `}</style>
    </React.Fragment>
  );
};

export default IconWrapper;
