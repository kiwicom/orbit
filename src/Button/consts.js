// @flow
export const TYPE_OPTIONS = {
  PRIMARY: "primary",
  SECONDARY: "secondary",
  INFO: "info",
  SUCCESS: "success",
  WARNING: "warning",
  CRITICAL: "critical",
  FACEBOOK: "facebook",
  GOOGLE: "google",
  WHITE: "white",
};

export const SIZE_OPTIONS = {
  SMALL: "small",
  NORMAL: "normal",
  LARGE: "large",
};

export const TOKENS = {
  // Size tokens
  heightButton: "heightButton",
  loadingWidth: "loadingWidth",
  loadingHeight: "loadingHeight",
  fontSizeButton: "fontSizeButton",
  paddingButton: "paddingButton",
  paddingButtonWithIcons: "paddingButtonWithIcons",
  paddingButtonWithLeftIcon: "paddingButtonWithLeftIcon",
  paddingButtonWithRightIcon: "paddingButtonWithRightIcon",
  marginRightIcon: "marginRightIcon",
  // Type tokens
  backgroundButton: "backgroundButton",
  backgroundButtonHover: "backgroundButtonHover",
  backgroundButtonActive: "backgroundButtonActive",
  backgroundButtonBordered: "backgroundButtonBordered",
  backgroundButtonBorderedHover: "backgroundButtonBorderedHover",
  backgroundButtonBorderedActive: "backgroundButtonBorderedActive",
  colorTextButton: "colorTextButton",
  colorTextButtonBordered: "colorTextButtonBordered",
  colorTextButtonHover: "colorTextButtonHover",
  colorTextButtonBorderedHover: "colorTextButtonBorderedHover",
  colorTextButtonActive: "colorTextButtonActive",
  colorTextButtonBorderedActive: "colorTextButtonBorderedActive",
  borderColorButton: "borderColorButton",
  borderColorButtonHover: "borderColorButtonHover",
  borderColorButtonActive: "borderColorButtonActive",
};

export const BUTTON_STATES = {
  DEFAULT: "default",
  HOVER: "hover",
  ACTIVE: "active",
  FOCUS: "focus",
};
