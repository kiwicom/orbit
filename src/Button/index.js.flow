// @flow
/*
  DOCUMENTATION: https://orbit.kiwi/components/button/
*/
import type { ReactComponentStyled } from "styled-components";

import type { Globals, Ref } from "../common/common.js.flow";

export type Type =
  | "primary"
  | "secondary"
  | "info"
  | "success"
  | "warning"
  | "critical"
  | "facebook"
  | "google"
  | "white";

export type Size = "small" | "normal" | "large";

export type Props = {|
  ...Globals,
  ...Ref,
  +children?: React$Node,
  +component?: string | React$Node,
  +href?: string,
  +onClick?: (e: SyntheticEvent<HTMLButtonElement>) => void | Promise<any>,
  +external?: boolean,
  +circled?: boolean,
  +bordered?: boolean,
  +disabled?: boolean,
  +block?: boolean,
  +loading?: boolean,
  +type?: Type,
  +size?: Size,
  +width?: number,
  +submit?: boolean,
  +icon?: React$Node,
  +iconLeft?: React$Node,
  +iconRight?: React$Node,
  +tabIndex?: string,
|};

declare export var StyledButton: ReactComponentStyled<Props>;

declare export default React$ComponentType<Props>;
