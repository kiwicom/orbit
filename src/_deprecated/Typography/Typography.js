// @flow
import * as React from "react";

import { fontColors } from "../../constants";

const sizes = {
  small: "12px",
  normal: "14px",
  large: "16px",
  header: "22px",
};

const additionalVariants = {
  bold: "font-weight: bold;",
  medium: "font-weight: 500;",
  normal: "",
};

type Props = {
  size?: $Keys<typeof sizes>,
  type?: "primary" | "secondary" | "attention" | "error" | "input" | "active" | "help",
  variant?: "bold" | "normal" | "medium",
  children: React.Node,
};

const Typography = (props: Props) => (
  <span>
    {props.children}
    <style jsx>{`
      span {
        font-family: Roboto, -apple-system, sans-serif;
        font-size: ${sizes[props.size || "normal"]};
        color: ${fontColors[props.type || "primary"]};
        ${additionalVariants[props.variant || "normal"]};
      }
    `}</style>
  </span>
);

export default Typography;
