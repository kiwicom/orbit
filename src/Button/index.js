// @flow
import * as React from "react";
import * as tokens from "orbit-design-token";
import { darken } from "polished";

import IconWrapper from "./IconWrapper";
import Loading from "../icons/Loading";

type Props = {
  title?: string,
  onClick: (e: SyntheticEvent<HTMLButtonElement>) => void,
  isDisabled?: boolean,
  isBordered?: boolean,
  isLoading?: boolean,
  isBlock?: boolean,
  type: "primary" | "secondary" | "link" | "facebook" | "google" | "destructive",
  size: "normal" | "large" | "small",
  width?: number,
  Icon?: React.ComponentType<*>,
  children?: React.Node,
};

const backgroundButton = {
  primary: tokens.backgroundButtonPrimary,
  secondary: tokens.backgroundButtonSecondary,
  link: tokens.backgroundButtonLink,
  facebook: tokens.backgroundButtonFacebook,
  google: tokens.backgroundButtonGoogle,
  destructive: tokens.backgroundButtonDestructive,
};

const borderButton = {
  primary: tokens.borderColorButtonPrimaryBordered,
  secondary: tokens.borderColorButtonSecondaryBordered,
  link: tokens.borderColorButtonLinkBordered,
  facebook: tokens.borderColorButtonFacebookBordered,
  google: tokens.borderColorButtonGoogleBordered,
  destructive: tokens.borderColorButtonDestructiveBordered,
};

const colorButton = {
  primary: tokens.colorTextButtonPrimary,
  secondary: tokens.colorTextButtonSecondary,
  link: tokens.colorTextButtonLink,
  facebook: tokens.colorTextButtonFacebook,
  google: tokens.colorTextButtonGoogle,
  destructive: tokens.colorTextButtonDestructive,
};

const colorButtonBordered = {
  primary: tokens.colorTextButtonPrimaryBordered,
  secondary: tokens.colorTextButtonSecondaryBordered,
  link: tokens.colorTextButtonLinkBordered,
  facebook: tokens.colorTextButtonFacebookBordered,
  google: tokens.colorTextButtonGoogleBordered,
  destructive: tokens.colorTextButtonDestructiveBordered,
};

const heightButton = {
  large: tokens.heightButtonLarge,
  normal: tokens.heightButtonNormal,
  small: tokens.heightButtonSmall,
};

const fontSizeButton = {
  large: tokens.fontSizeButtonLarge,
  normal: tokens.fontSizeButtonNormal,
  small: tokens.fontSizeButtonSmall,
};

const paddingButton = {
  large: tokens.paddingButtonLarge,
  normal: tokens.paddingButtonNormal,
  small: tokens.paddingButtonSmall,
};

const paddingButtonWithIcon = {
  large: tokens.paddingButtonLargeWithIcon,
  normal: tokens.paddingButtonNormalWithIcon,
  small: tokens.paddingButtonSmallWithIcon,
};

const Button = (props: Props) => {
  let widthCondition;
  let backgroundHoverCondition;
  let backgroundActiveCondition;

  if (props.isBlock) {
    widthCondition = "100%";
  } else if (props.width) {
    widthCondition = `${props.width}px`;
  } else {
    widthCondition = "auto";
  }

  if (props.isBordered) {
    backgroundHoverCondition = tokens.backgroundButtonBordered;
  } else if (props.type === "link") {
    backgroundHoverCondition = tokens.backgroundButtonLinkHover;
  } else {
    backgroundHoverCondition = darken(tokens.modifierDarkenHover, backgroundButton[props.type]);
  }

  if (props.isBordered) {
    backgroundActiveCondition = tokens.backgroundButtonBordered;
  } else if (props.type === "link") {
    backgroundActiveCondition = tokens.backgroundButtonLinkHover;
  } else {
    backgroundActiveCondition = darken(tokens.modifierDarkenActive, backgroundButton[props.type]);
  }
  return (
    <button onClick={props.onClick} disabled={props.isDisabled}>
      {props.Icon && (
        <IconWrapper
          Icon={props.isLoading ? Loading : props.Icon}
          size={props.size}
          type={props.type}
          bordered={props.isBordered}
        />
      )}
      {props.children ? props.children : props.title}
      <style jsx>{`
        button {
          box-sizing: border-box;
          display: flex;
          justify-content: center;
          align-items: center;
          width: ${widthCondition};
          min-width: ${tokens.widthButtonMinimal};
          height: ${heightButton[props.size]};
          background: ${props.isBordered
            ? tokens.backgroundButtonBordered
            : backgroundButton[props.type]};
          color: ${props.isBordered ? colorButtonBordered[props.type] : colorButton[props.type]};
          border: ${props.isBordered ? `1px solid ${borderButton[props.type]}` : "none"};
          border-radius: ${tokens.borderRadiusNormal};
          padding-left: ${props.Icon
            ? paddingButtonWithIcon[props.size]
            : paddingButton[props.size]};
          padding-right: ${paddingButton[props.size]};
          font-weight: ${tokens.fontWeightBold};
          font-size: ${fontSizeButton[props.size]};
          cursor: ${props.isDisabled ? "default" : "pointer"};
          opacity: ${props.isDisabled ? tokens.opacityButtonDisabled : "1"};
          transition: all 0.15s ease-in-out;
          outline: 0;
        }
        button:hover {
          background: ${backgroundHoverCondition};
          color: ${props.isBordered
            ? darken(tokens.modifierDarkenHover, colorButtonBordered[props.type])
            : colorButton[props.type]};
          border: ${props.isBordered
            ? `1px solid ${darken(tokens.modifierDarkenHover, borderButton[props.type])}`
            : "none"};
        }
        button:active {
          transform: scale(${tokens.modifierScaleButtonActive});
          background: ${backgroundActiveCondition};
          color: ${props.isBordered
            ? darken(tokens.modifierDarkenActive, colorButtonBordered[props.type])
            : colorButton[props.type]};
          border: ${props.isBordered
            ? `1px solid ${darken(tokens.modifierDarkenActive, borderButton[props.type])}`
            : "none"};
        }
      `}</style>
    </button>
  );
};

Button.defaultProps = {
  isDisabled: false,
  isBordered: false,
  isLoading: false,
  isBlock: false,
  width: 180,
};

export default Button;
