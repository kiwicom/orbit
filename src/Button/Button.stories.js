// @flow
import * as React from "react";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import { text, number, boolean, select } from "@storybook/addon-knobs";

import * as Icons from "../icons";
import { TYPE_OPTIONS, SIZE_OPTIONS } from "./consts";
import RenderInRtl from "../utils/rtl/RenderInRtl";
import Stack from "../Stack";
import SPACINGS_AFTER from "../common/getSpacingToken/consts";

import Button from "./index";

const getIcons = (name, defaultIcon) => select(name, [null, ...Object.keys(Icons)], defaultIcon);
const getIcon = source => Icons[source];

storiesOf("Button", module)
  .add(
    "Default",
    () => {
      const children = text("Children", "Default button");
      return <Button onClick={action("clicked")}>{children}</Button>;
    },
    {
      info:
        "This is the default configuration of this component. Visit Orbit.Kiwi for more detailed guidelines.",
    },
  )
  .add(
    "Basic buttons",
    () => {
      const children = text("Children", "Button");
      const fullWidth = boolean("fullWidth", false);
      const type = select("Type", [TYPE_OPTIONS.PRIMARY, TYPE_OPTIONS.SECONDARY], "primary");
      const size = select("Size", Object.values(SIZE_OPTIONS), "normal");

      return (
        <Button onClick={action("clicked")} fullWidth={fullWidth} type={type} size={size}>
          {children}
        </Button>
      );
    },
    {
      info:
        "Basic buttons have three sizes (large, normal and small) and can be either primary or secondary type. Visit Orbit.Kiwi for more detailed guidelines.",
    },
  )
  .add(
    "Button with icons",
    () => {
      const children = text("Children", "Button");
      const fullWidth = boolean("fullWidth", false);
      const type = select("Type", [TYPE_OPTIONS.PRIMARY, TYPE_OPTIONS.SECONDARY], "primary");
      const size = select("Size", Object.values(SIZE_OPTIONS), "small");
      const IconLeft = getIcon(getIcons("iconLeft", "PlusCircle"));
      const IconRight = getIcon(getIcons("iconRight", "ChevronDown"));
      return (
        <Button
          onClick={action("clicked")}
          fullWidth={fullWidth}
          type={type}
          size={size}
          iconLeft={IconLeft && <IconLeft />}
          iconRight={IconRight && <IconRight />}
        >
          {children}
        </Button>
      );
    },
    {
      info:
        "Buttons with icon are great when you need to draw more attention to the action. However, it's essential to not over-use these buttons. If everything tries to grab attention, things usually get messy. Visit Orbit.Kiwi for more detailed guidelines.",
    },
  )
  .add(
    "Facebook button",
    () => (
      <Button onClick={action("clicked")} type="facebook" iconLeft={<Icons.Facebook />} bordered>
        Sign in with Facebook
      </Button>
    ),
    {
      info: "We use social buttons only in normal size.",
    },
  )
  .add(
    "Google button",
    () => (
      <Button onClick={action("clicked")} type="google" iconLeft={<Icons.Google />} bordered>
        Sign in with Google
      </Button>
    ),
    {
      info: "We use social buttons only in normal size.",
    },
  )
  .add(
    "Status buttons",
    () => {
      const children = text("Children", "Button");
      const IconLeft = getIcon(getIcons("iconLeft", "CloseCircle"));

      return (
        <Stack>
          <Button
            onClick={action("clicked")}
            type="info"
            size="small"
            iconLeft={IconLeft && <IconLeft />}
          >
            {children}
          </Button>
          <Button
            onClick={action("clicked")}
            type="success"
            size="small"
            iconLeft={IconLeft && <IconLeft />}
          >
            {children}
          </Button>
          <Button
            onClick={action("clicked")}
            type="warning"
            size="small"
            iconLeft={IconLeft && <IconLeft />}
          >
            {children}
          </Button>
          <Button
            onClick={action("clicked")}
            type="critical"
            size="small"
            iconLeft={IconLeft && <IconLeft />}
          >
            {children}
          </Button>
        </Stack>
      );
    },
    {
      info:
        "We use status buttons exclusively in Alert messages when we need to show supporting action connected to the displayed message. We only use the small size of buttons. Visit Orbit.Kiwi for more detailed guidelines.",
    },
  )
  .add(
    "Circled button",
    () => {
      const size = select("Size", Object.values(SIZE_OPTIONS), SIZE_OPTIONS.NORMAL);
      const type = select(
        "Type",
        [TYPE_OPTIONS.PRIMARY, TYPE_OPTIONS.SECONDARY],
        TYPE_OPTIONS.PRIMARY,
      );
      const IconLeft = getIcon(getIcons("iconLeft", "Airplane"));

      return (
        <Button
          onClick={action("clicked")}
          size={size}
          type={type}
          iconLeft={IconLeft && <IconLeft />}
          circled
          title="Button"
        />
      );
    },
    {
      info:
        "Button can be also rendered in circled shape, but only when it does have iconLeft and not have children. Use title prop to ensure accessibility.",
    },
  )
  .add(
    "Destructive buttons",
    () => {
      const children = text("Children", "Destructive button");
      const bordered = boolean("Bordered", false);
      const size = select("Size", Object.values(SIZE_OPTIONS), "normal");

      return (
        <Button
          onClick={action("clicked")}
          bordered={bordered}
          type="critical"
          size={size}
          iconLeft={<Icons.Remove />}
        >
          {children}
        </Button>
      );
    },
    {
      info:
        "Destructive buttons are a specific version of critical status buttons, paired together with 'Remove' icon. We use them when we need to inform our users about possible dangerous actions (canceling a booking, removing an item, etc.). Visit Orbit.Kiwi for more detailed guidelines.",
    },
  )
  .add(
    "Button as a link",
    () => {
      const children = text("Children", "I am a link");
      const href = text("Href", "https://kiwi.com");
      const external = boolean("External", false);
      const disabled = boolean("Disabled", false);
      const size = select("Size", Object.values(SIZE_OPTIONS), "normal");

      return (
        <Button
          onClick={action("clicked")}
          href={href}
          external={external}
          size={size}
          disabled={disabled}
          bordered
          iconLeft={<Icons.Airplane />}
        >
          {children}
        </Button>
      );
    },
    {
      info:
        "If you need to, you can pass some href to this component and it will automatically render into anchor.",
    },
  )
  .add(
    "Playground",
    () => {
      const children = text("Children", "Button");
      const href = text("Href", undefined);
      const external = boolean("External", false);
      const asComponent = text("asComponent", "button");
      const disabled = boolean("Disabled", false);
      const fullWidth = boolean("fullWidth", false);
      const type = select("Type", Object.values(TYPE_OPTIONS), "primary");
      const size = select("Size", Object.values(SIZE_OPTIONS), "normal");
      const width = number("Width", 0);
      const bordered = boolean("Bordered", false);
      const circled = boolean("Circled", false);
      const loading = boolean("Loading", false);
      const submit = boolean("Submit", false);
      const dataTest = text("dataTest", "test");
      const IconLeft = getIcon(getIcons("iconLeft", "Airplane"));
      const IconRight = getIcon(getIcons("iconRight", "ChevronDown"));
      const ariaExpanded = boolean("Aria expanded", false);
      const ariaControls = text("Aria controls", "element ID");
      const tabIndex = text("tabIndex", "0");
      const spaceAfter = select("spaceAfter", [null, ...Object.values(SPACINGS_AFTER)]);
      const title = text("Title", "Additional information for accessibility");

      return (
        <Button
          onClick={action("clicked")}
          asComponent={asComponent}
          href={href}
          external={external}
          disabled={disabled}
          circled={circled}
          fullWidth={fullWidth}
          bordered={bordered}
          loading={loading}
          dataTest={dataTest}
          type={type}
          size={size}
          iconLeft={IconLeft && <IconLeft />}
          iconRight={IconRight && <IconRight />}
          submit={submit}
          width={width}
          ariaControls={ariaControls}
          ariaExpanded={ariaExpanded}
          tabIndex={tabIndex}
          spaceAfter={spaceAfter}
          title={title}
        >
          {children}
        </Button>
      );
    },
    {
      info: "Some description about this type of component. ",
    },
  )
  .add(
    "Accessibility",
    () => {
      const children = text("Children", "Button");
      const ariaExpanded = boolean("Aria expanded", false);
      const ariaControls = text("Aria controls", "element ID");
      const title = text("Title", "Additional information for accessibility");

      return (
        <Button
          onClick={action("clicked")}
          ariaControls={ariaControls}
          ariaExpanded={ariaExpanded}
          title={title}
        >
          {children}
        </Button>
      );
    },
    {
      info: "This is a preview of component accessibility props",
    },
  )
  .add(
    "RTL",
    () => (
      <RenderInRtl>
        <Button type="info" icon={<Icons.Airplane />}>
          Button
        </Button>
      </RenderInRtl>
    ),
    {
      info: "This is a preview of this component in RTL setup.",
    },
  );
