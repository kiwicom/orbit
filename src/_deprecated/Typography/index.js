// @flow
import Typography from "./Typography";
import DeprecatedComponent from "../DeprecatedComponent";

export default DeprecatedComponent(
  Typography,
  "Typography is deprecated and will be removed in next major release",
);
