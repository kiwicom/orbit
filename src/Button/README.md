# Button

To implement Button component into your project you'll need to add the import:

```jsx
import Button from "@kiwicom/orbit-components/lib/Button";
```

After adding import into your project you can use it simply like:

```jsx
<Button>Hello World!</Button>
```

## Props

Table below contains all types of the props available in Button component.

| Name          | Type                       | Default     | Description                                                                                                                                     |
| :------------ | :------------------------- | :---------- | :---------------------------------------------------------------------------------------------------------------------------------------------- |
| ariaControls  | `string`                   |             | Id of the element the button controls.                                                                                                          |
| ariaExpanded  | `boolean`                  |             | Tells screen reader the controlled element from `ariaControls` is expanded                                                                      |
| bordered      | `boolean`                  | `false`     | If `true`, the Button will have a lighter version, with border and light background.                                                            |
| circled       | `boolean`                  | `false`     | If `true`, the Button will have circular shape.                                                                                                 |
| children      | `React.Node`               |             | The content of the Button. [See Functional specs](#functional-specs)                                                                            |
| **component** | `string \| React.Node`     | `"button"`  | The component used for the root node. Either a string to use a DOM element or a component.                                                      |
| dataTest      | `string`                   |             | Optional prop for testing purposes.                                                                                                             |
| disabled      | `boolean`                  | `false`     | If `true`, the Button will be disabled.                                                                                                         |
| external      | `boolean`                  | `false`     | If `true`, the Button opens link in a new tab. [See Functional specs](#functional-specs)                                                        |
| fullWidth     | `boolean`                  | `false`     | If `true`, the Button will grow up to the full width of its container.                                                                          |
| href          | `string`                   |             | The URL of the link to open when Button is clicked. [See Functional specs](#functional-specs)                                                   |
| icon          | `React.Node`               |             | The displayed icon (will be removed in the future, use iconLeft instead).                                                                       |
| iconLeft      | `React.Node`               |             | The displayed icon on the left.                                                                                                                 |
| iconRight     | `React.Node`               |             | The displayed icon on the right.                                                                                                                |
| loading       | `boolean`                  | `false`     | If `true`, the loading glyph will be displayed.                                                                                                 |
| onClick       | `event => void \| Promise` |             | Function for handling onClick event.                                                                                                            |
| ref           | `func`                     |             | Prop for forwarded ref of the Button.                                                                                                           |
| role          | `string`                   |             | Specifies the role of an element.                                                                                                               |
| **size**      | [`enum`](#enum)            | `"normal"`  | The size of the Button.                                                                                                                         |
| spaceAfter    | `enum`                     |             | Additional `margin-bottom` after component. [See this docs](https://github.com/kiwicom/orbit-components/tree/master/src/common/getSpacingToken) |
| submit        | `boolean`                  | `false`     | If `true`, the Button will have `type="submit"` attribute, otherwise `type="button"`.                                                           |
| tabIndex      | `string`                   |             | Specifies the tab order of an element.                                                                                                          |
| title         | `string`                   |             | Adds `aria-label`.                                                                                                                              |
| **type**      | [`enum`](#enum)            | `"primary"` | The type of Button.                                                                                                                             |
| width         | `number`                   | `0`         | The width of the Button. Number is defined in `px`.                                                                                             |

### enum

| type          | size       |
| :------------ | :--------- |
| `"primary"`   | `"small"`  |
| `"secondary"` | `"normal"` |
| `"info"`      | `"large"`  |
| `"success"`   |            |
| `"warning"`   |            |
| `"critical"`  |            |
| `"facebook"`  |            |
| `"google"`    |            |
| `"white"`     |            |

## Functional specs

- When the `external` is specified, `noopener` and `noreferrer` values will automatically added to attribute `rel` for security reason.

- By passing the `href` prop into Button, it will render into `a` element. If you pass `component` prop it will override this logic.

- If you want to render **Icon only Button**, you just need to let `children` prop empty and set up any `icon` you want to use.

- If you want to use the `component` prop then **YourComponent** should accept at least `className`. Otherwise it won't be rendered with proper styling, e.g.:

  ```jsx
  const YourComponent = props => <div {...props} />

  <Button component={YourComponent}>Title</Button>
  ```

  If you specify the children of **YourComponent** component, it will override the children prop of Button component, e.g.:

  ```jsx
  const YourComponent = props => <div {...props}>YourComponent</div>;
  ```

## Accessibility

A button is mainly used for indicating an action, submitting a data, opening a modal etc. If you want to use Button for navigation consider using a `<TextLink>` for that.

- Use `ariaControls` prop to add `aria-controls` attribute to establish the relationship between button and element which is controlled by it. `aria-controls` works only with a unique `id` of an element.

- Use `ariaExpands` prop to add `aria-expands` to indicate screenreaders, that element controlled by button through `ariaControls` is expanded or not.

- Use `disabled` prop to indicate users that button is inactive and they can't interact with it.

- Use `role` and `tabIndex` when you are rendering `Button` to non-actionable HTML element as `div` or `span`. However, this should be done only in edge-cases as it is anti-pattern behavior.

- Use `title` to add `aria-label` when you need to add extra informations to screen readers or there is no `children` presented to be used as label.
