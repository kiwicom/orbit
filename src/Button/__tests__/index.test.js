// @flow

import * as React from "react";
import { shallow } from "enzyme";

import Button from "../";
import Typography from "../../Typography";
import Airplane from "../../icons/Airplane";

describe("Button", () => {
  const title = "Title";
  const component = shallow(
    <Button title={title} size="normal" type="primary" onClick={jest.fn()} />,
  );
  const button = component.find("button");
  it("Should contain a title ", () => {
    expect(button.render().text()).toBe(title);
  });
});

describe("When button is clicked", () => {
  const onClick = jest.fn();
  const title = "title";
  const component = shallow(
    <Button type="primary" size="normal" title={title} onClick={onClick} />,
  );
  const button = component.find("button");

  it("should execute onClick method", () => {
    button.simulate("click");
    expect(onClick).toHaveBeenCalled();
  });
});

describe("Rendered with icon", () => {
  const title = "title";
  const icon = () => <i />;
  const component = shallow(
    <Button type="primary" size="normal" title={title} Icon={icon} onClick={jest.fn()} />,
  );
  const button = component.find("button");
  it("Should contain IconWrapper", () => {
    expect(button.find("IconWrapper").exists()).toBe(true);
  });
});

describe("Rendered with children", () => {
  const component = shallow(
    <Button type="primary" onClick={jest.fn()} size="large">
      <Typography type="primary" variant="bold">
        Typo children
      </Typography>
      <Airplane fill="#F2473F" size="32" />
    </Button>,
  );
  it("Should contain IconWrapper", () => {
    expect(component).toMatchSnapshot();
  });
});
