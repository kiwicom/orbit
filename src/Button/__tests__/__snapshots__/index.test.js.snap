// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendered with children Should contain IconWrapper 1`] = `
<button
  className="jsx-588294754"
  disabled={false}
  onClick={[MockFunction]}
>
  <Typography
    type="primary"
    variant="bold"
  >
    Typo children
  </Typography>
  <Airplane
    fill="#F2473F"
    size="32"
  />
  <JSXStyle
    css="button.__jsx-style-dynamic-selector{box-sizing:border-box;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-pack:center;-webkit-justify-content:center;-ms-flex-pack:center;justify-content:center;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;width:180px;min-width:180px;height:52px;background:#00a991;color:#fff;border:none;border-radius:3px;padding-left:28px;padding-right:28px;font-weight:700;font-size:16px;cursor:pointer;opacity:1;-webkit-transition:all 0.15s ease-in-out;transition:all 0.15s ease-in-out;outline:0;}button.__jsx-style-dynamic-selector:hover{background:#00907b;color:#fff;border:none;}button.__jsx-style-dynamic-selector:active{-webkit-transform:scale(0.9);-ms-transform:scale(0.9);transform:scale(0.9);background:#007665;color:#fff;border:none;}"
    dynamic={
      Array [
        "180px",
        "180px",
        "52px",
        "#00a991",
        "#fff",
        "none",
        "3px",
        "28px",
        "28px",
        "700",
        "16px",
        "pointer",
        "1",
        "#00907b",
        "#fff",
        "none",
        0.9,
        "#007665",
        "#fff",
        "none",
      ]
    }
    styleId="4169747830"
  />
</button>
`;
